examples/step-62/step-62.cc

```diff
@@ -0,0 +1,1468 @@
+    // constructor.
+    const double pml_coeff;
+    const int    pml_coeff_degree;
+    const double dimension_x;
+    const double dimension_y;
+    const bool   pml_x;
+    const bool   pml_y;
+    const double pml_width_x;
+    const double pml_width_y;
+    const double a_coeff_x;
+    const double a_coeff_y;
+  };
+
+
+
+  // @sect4{`Rho` class}
+  // This class is used to define the mass density.
+  template <int dim>
+  class Rho : public Function<dim>
+  {
+  public:
+    Rho(HDF5::Group &data);
+    virtual double value(const Point<dim> & p,
+                         const unsigned int component = 0) const;
+
+  private:
+    // HDF5::Group in which all the simulation results will be stored.
+    HDF5::Group data;
+
+    // The same as before, the following attributes are defined in the jupyter
+    // notebook, stored in `data` as HDF5 attributes and then read by the
+    // constructor.
+    const double       lambda;
+    const double       mu;
+    const double       material_a_rho;
+    const double       material_b_rho;
+    const double       cavity_resonance_frequency;
+    const unsigned int nb_mirror_pairs;
+    const double       dimension_y;
+    const unsigned int grid_level;
+    double             average_rho_width;
+  };
+
+
+
+  // @sect4{`Parameters` class}
+  // This class contains all the parameters that will be used in the simulation.
+  template <int dim>
+  class Parameters
+  {
+  public:
+    Parameters(HDF5::Group &data);
+
+    // HDF5::Group in which all the simulation results will be stored.
+    HDF5::Group data;
+
+    // The same as before, the following attributes are defined in the jupyter
+    // notebook, stored in `data` as HDF5 attributes and then read by the
+    // constructor.
+    const std::string        simulation_name;
+    bool                     save_vtu_files;
+    const double             start_frequency;
+    const double             stop_frequency;
+    const unsigned int       nb_frequency_points;
+    const double             lambda;
+    const double             mu;
+    const double             dimension_x;
+    const double             dimension_y;
+    const unsigned int       nb_probe_points;
+    const unsigned int       grid_level;
+    Point<dim>               probe_start_point;
+    Point<dim>               probe_stop_point;
+    const RightHandSide<dim> right_hand_side;
+    const PML<dim>           pml;
+    const Rho<dim>           rho;
+
+  private:
+    const double comparison_float_constant = 1e-12;
+  };
+
+
+
+  // @sect4{`PointHistory` class}
+  // The calculation of the mass and stiffness matrices is very expensive. These
+  // matrices are the same for all the frequency steps. The right hand side
+  // vector is also the same for all the frequency steps. We use this class to
+  // store these values and re-use them at each frequency step. The
+  // `PointHistory` class has already been used in step-18.
+
+  template <int dim>
+  class PointHistory
```

Comment:
It's not clear to me why this class is called `PointHistory`. You're just caching data. What is the connection?

Also, I believe that what you store here are not the *assembled* matrices and right hand sides, but instead the data for a single cell. Is this right? Maybe you could put that into the documentation of the class?

Reply:
I agree PointHistory is not a good name. Now the class is named `QuadratureCache`. Is that a good name?

Reply:
Yes!